### PR TITLE
Add extended mnemonics

### PR DESCRIPTION
**Pull request by dominik-steenken**
This PR adds a number of thus-far missing extended mnemonics to the assembler and disassembler for SystemZ.

The following mnemonics have been added and are supported for the assembler and disassembler:

- `NOP(R)?`
- `LFI`
- `RISBG(N)?Z`

The following mnemonics have been added and are supported for the assembler only:

- `JC(TH)?`
- `LLG(F|H)I`
- `NOT(G)?R`

### CODE REVIEW

File: llvm/lib/Target/SystemZ/SystemZInstrInfo.td [hunk 1]
```diff
@@ -973,6 +980,7 @@ let isAsCheapAsAMove = 1, isMoveImm = 1, isReMaterializable = 1 in {
   def IILF : UnaryRIL<"iilf", 0xC09, bitconvert, GR32, uimm32>;
   def IIHF : UnaryRIL<"iihf", 0xC08, bitconvert, GRH32, uimm32>;
 }
+def LFI : InstAlias<"lfi\t$R1, $RI1",(IILF GR32:$R1, uimm32:$RI1)>;
```
Comment (redstar) [inline llvm/lib/Target/SystemZ/SystemZInstrInfo.td:983]:
```suggestion
def LFI : InstAlias<"lfi\t$R1, $RI1",(IILF GR32:$R1, uimm32:$RI1)>;
```
```suggestion
def LFI : InstAlias<"lfi\t$R1, $RI1", (IILF GR32:$R1, uimm32:$RI1)>;
```

File: llvm/lib/Target/SystemZ/SystemZInstrInfo.td [hunk 2]
```diff
@@ -464,6 +469,8 @@ let isAsCheapAsAMove = 1, isMoveImm = 1, isReMaterializable = 1 in {
   def LLILF : UnaryRIL<"llilf", 0xC0F, bitconvert, GR64, imm64lf32>;
   def LLIHF : UnaryRIL<"llihf", 0xC0E, bitconvert, GR64, imm64hf32>;
 }
+def LLGFI : InstAlias<"llgfi\t$R1, $RI1",(LLILF GR64:$R1, imm64lf32:$RI1)>;
+def LLGHI : InstAlias<"llghi\t$R1, $RI1",(LLILL GR64:$R1, imm64ll16:$RI1)>;
```
Comment (redstar) [inline llvm/lib/Target/SystemZ/SystemZInstrInfo.td:473]:
```suggestion
def LLGFI : InstAlias<"llgfi\t$R1, $RI1", (LLILF GR64:$R1, imm64lf32:$RI1)>;
def LLGHI : InstAlias<"llghi\t$R1, $RI1", (LLILL GR64:$R1, imm64ll16:$RI1)>;
```
```suggestion
def LLGFI : InstAlias<"llgfi\t$R1, $RI1",(LLILF GR64:$R1, imm64lf32:$RI1)>;
def LLGHI : InstAlias<"llghi\t$R1, $RI1",(LLILL GR64:$R1, imm64ll16:$RI1)>;
```

File: llvm/lib/Target/SystemZ/SystemZInstrInfo.td [hunk 3]
```diff
@@ -111,11 +111,16 @@ let isBranch = 1, isTerminator = 1, isBarrier = 1 in {
...
   def NOP_bare  : InstRXb<0x47,(outs), (ins), "nop", []>;
-def NOPR : InstAlias<"nopr\t$R", (BCRAsm 0, GR64:$R), 0>;
-def NOPR_bare : InstAlias<"nopr", (BCRAsm 0, R0D), 0>;
+def NOPR : NeverCondBranchRR<"nopr", 0x07>;
+def NOPR_bare : InstAlias<"nopr", (NOPR R0D), 0>;
+// def NOP  : InstAlias<"nop\t$XBD", (BCAsm 0, bdxaddr12only:$XBD), 0>;
+// let isAsmParserOnly = 1, hasNoSchedulingInfo = 1, M1 = 0, X2 = 0, B2 = 0, D2 = 0 in
+//   def NOP_bare  : InstRXb<0x47,(outs), (ins), "nop", []>;
+// def NOPR : InstAlias<"nopr\t$R", (BCRAsm 0, GR64:$R), 0>;
+// def NOPR_bare : InstAlias<"nopr", (BCRAsm 0, R0D), 0>;
```
Comment (redstar) [inline llvm/lib/Target/SystemZ/SystemZInstrInfo.td:123]: No dead code in comments please.

File: llvm/lib/Target/SystemZ/SystemZInstrFormats.td [hunk 4]
```diff
@@ -363,7 +363,7 @@ class InstRIEe<bits<16> op, dag outs, dag ins, string asmstr, list<dag> pattern>
   let Inst{7-0}   = op{7-0};
 }
 
-class InstRIEf<bits<16> op, dag outs, dag ins, string asmstr, list<dag> pattern>
+class InstRIEf<bits<16> op, dag outs, dag ins, string asmstr, list<dag> pattern, bits<8> I3Or = 0, bits<8> I4Or = 0>
```
Comment (redstar) [inline llvm/lib/Target/SystemZ/SystemZInstrFormats.td:366]: Not sure how serious this is applied to .td files, but the line is way over the 80 char limit. (Well, the original line is a little bit too long, too.)

**Review by redstar — COMMENTED** [review]
Small style issue.

File: llvm/lib/Target/SystemZ/SystemZInstrInfo.td (hunk 1 above)
Comment (dominik-steenken) [reply llvm/lib/Target/SystemZ/SystemZInstrInfo.td]: fixed

File: llvm/lib/Target/SystemZ/SystemZInstrInfo.td (hunk 2 above)
Comment (dominik-steenken) [reply llvm/lib/Target/SystemZ/SystemZInstrInfo.td]: fixed

File: llvm/lib/Target/SystemZ/SystemZInstrInfo.td (hunk 3 above)
Comment (dominik-steenken) [reply llvm/lib/Target/SystemZ/SystemZInstrInfo.td]: goddamnit. removed.

File: llvm/lib/Target/SystemZ/SystemZInstrFormats.td (hunk 4 above)
Comment (dominik-steenken) [reply llvm/lib/Target/SystemZ/SystemZInstrFormats.td]: I split the line, similar to how it is done in the handful of other formats with extra parms.

**Comment by uweigand** [comment]
The CI shows failures in insn-good.s and a bunch of SystemZ LLD tests, these definitely need to be addressed.

Also, can you re-check that all mnemonics added here are also accepted by GAS?  I think at least these may not be:  nhhr, nhlr, nlhr, ohhr, ohlr, olhr, xhhr, xhlr, xlhr.   You should verify that GAS can assemble all the insn-good* files (resulting in mostly identical object files - at least the disassembly of those object files needs to match again).

**Comment by dominik-steenken** [comment]
> The CI shows failures in insn-good.s and a bunch of SystemZ LLD tests, these definitely need to be addressed.

I have fixed those tests. In `insn-good.s` some Aliases were pinned to the wrong target instruction, and the *lld* tests needed to be modified to account for `nop` and `nopr` now being emitted by the disassembler.

> Also, can you re-check that all mnemonics added here are also accepted by GAS? I think at least these may not be: nhhr, nhlr, nlhr, ohhr, ohlr, olhr, xhhr, xhlr, xlhr. You should verify that GAS can assemble all the insn-good* files (resulting in mostly identical object files - at least the disassembly of those object files needs to match again).

I moved the extended mnemonics of `R(N|O|X)SBG` to a different PR and re-checked that gas can assemble all of our assembly tests. While doingf that, we discovered an issue with register types in some instructions, as well as a difference in permissiveness regarding the immediate fields with `Z` and `T` bits that will be addressed in separate PRs.

**Review by uweigand — APPROVED** [review]
LGTM now, thanks!

**Comment by dominik-steenken** [comment]
follow-up build to the above is clean

**Pull request merged**